### PR TITLE
[Connectivity] migrate to the new android embedding

### PR DESCRIPTION
**Pull request by cyanglaz**
## Description

Migrate connectivity to the new android embedding.

## Related Issues

https://github.com/flutter/flutter/issues/41835

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

File: packages/connectivity/example/android/app/build.gradle [hunk 1]
```diff
@@ -52,6 +52,9 @@ flutter {
 }
 
 dependencies {
+    implementation 'androidx.lifecycle:lifecycle-runtime:2.1.0'
```
Comment (amirh) [inline packages/connectivity/example/android/app/build.gradle]: remove these deps, the engine POM makes sure these are pulled in (once https://github.com/flutter/flutter/pull/41885 lands)

File: packages/connectivity/android/src/main/java/io/flutter/plugins/connectivity/ConnectivityMethodChannelHandler.java [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+  private void handleBSSID(MethodCall call, MethodChannel.Result result) {
+    WifiInfo wifiInfo = getWifiInfo();
+    String bssid = null;
+    if (wifiInfo != null) bssid = wifiInfo.getBSSID();
+    result.success(bssid);
+  }
+
+  private void handleWifiIPAddress(MethodCall call, final MethodChannel.Result result) {
+    WifiManager wifiManager =
```
Comment (matthew-carroll) [inline packages/connectivity/android/src/main/java/io/flutter/plugins/connectivity/ConnectivityMethodChannelHandler.java]: Recommendation: passing this as a constructor argument allows you to mock for tests.

File: packages/connectivity/android/src/main/java/io/flutter/plugins/connectivity/ConnectivityMethodChannelHandler.java [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+      case ConnectivityManager.TYPE_MOBILE:
+      case ConnectivityManager.TYPE_MOBILE_DUN:
+      case ConnectivityManager.TYPE_MOBILE_HIPRI:
+        return "mobile";
+      default:
+        return "none";
+    }
+  }
+
+  private BroadcastReceiver createReceiver(final EventChannel.EventSink events) {
```
Comment (matthew-carroll) [inline packages/connectivity/android/src/main/java/io/flutter/plugins/connectivity/ConnectivityMethodChannelHandler.java]: Recommendation: If I were migrating this, I would probably restructure such that this `BroadcastReceiver` is separated from this `MethodCallHandler` and then bring them together in a third class.

File: packages/connectivity/android/src/main/java/io/flutter/plugins/connectivity/ConnectivityMethodChannelHandler.java [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+  public ConnectivityMethodChannelHandler(Context context) {
+    assert (context != null);
+    this.context = context;
+    this.manager =
+        (ConnectivityManager)
+            context.getApplicationContext().getSystemService(Context.CONNECTIVITY_SERVICE);
+  }
+
+  @Override
+  public void onListen(Object arguments, EventChannel.EventSink events) {
```
Comment (matthew-carroll) [inline packages/connectivity/android/src/main/java/io/flutter/plugins/connectivity/ConnectivityMethodChannelHandler.java]: Please add nullability annotations to all input parameters and return values for each public and private method.

File: packages/connectivity/android/src/main/java/io/flutter/plugins/connectivity/ConnectivityMethodChannelHandler.java [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+  }
+
+  @Override
+  public void onCancel(Object arguments) {
+    context.unregisterReceiver(receiver);
+    receiver = null;
+  }
+
+  @Override
+  public void onMethodCall(MethodCall call, MethodChannel.Result result) {
```
Comment (matthew-carroll) [inline packages/connectivity/android/src/main/java/io/flutter/plugins/connectivity/ConnectivityMethodChannelHandler.java]: Would you mind adding unit tests for this class to ensure that incoming messages are correctly parsed, and outgoing messages are correctly serialized?

File: packages/connectivity/android/src/main/java/dev/flutter/plugins/connectivity/ConnectivityPlugin.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+// found in the LICENSE file.
+
+package dev.flutter.plugins.connectivity;
+
+import io.flutter.embedding.engine.plugins.FlutterPlugin;
+import io.flutter.plugin.common.EventChannel;
+import io.flutter.plugin.common.MethodChannel;
+import io.flutter.plugins.connectivity.ConnectivityMethodChannelHandler;
+
+public class ConnectivityPlugin implements FlutterPlugin {
```
[the remark this answers is not in the document]
Comment (amirh) [reply packages/connectivity/android/src/main/java/dev/flutter/plugins/connectivity/ConnectivityPlugin.java]: New classes members/methods added should definitely be added with Javadocs.

Regarding all existing code - this is definitely tech debt we should be paying, and is welcomed if the author is interested in doing it as part of this PR, though it shouldn't block it. For the purpose of supporting the v2 embedder in all plugins I suggest we focus PRs and reviews on that goal. Please do leave any improvement suggestions of existing code as non-blocking nits.

I think this fits with the spirit of the [Flutter code review guidelines](https://github.com/flutter/flutter/wiki/Tree-hygiene#how-to-review-code):

> In general, reviewers should favor approving a PR once it is in a state where it definitely improves the overall code health of the system being worked on, even if the PR isn't perfect.

**Comment by cyanglaz** [comment]
@matthew-carroll Updated the code per some of our recommendation. I didn't add any JUnit tests nor Nullability notation to the existing code. Could you take another look and let me know if the new refactoring looks good.

File: packages/connectivity/android/src/main/java/io/flutter/plugins/connectivity/ConnectivityMethodChannelHandler.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+import io.flutter.plugin.common.MethodCall;
+import io.flutter.plugin.common.MethodChannel;
+
+/** Handles MethodChannel for the plugin. */
+public class ConnectivityMethodChannelHandler implements MethodChannel.MethodCallHandler {
+
+  private Connectivity connectivity;
+
+  /**
+   * Construct the ConnectivityMethodChannelHandler
```
Comment (matthew-carroll) [inline packages/connectivity/android/src/main/java/io/flutter/plugins/connectivity/ConnectivityMethodChannelHandler.java]: Probably "useless documentation"

File: packages/connectivity/android/src/main/java/io/flutter/plugins/connectivity/BroadcastReceiverRegistrarImpl.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+import androidx.annotation.NonNull;
+import io.flutter.plugin.common.EventChannel;
+
+/** The BroadcastReceiverRegistrar used for the plugin. */
+public class BroadcastReceiverRegistrarImpl implements BroadcastReceiverRegistrar {
+  private Context context;
+  private Connectivity connectivity;
+
+  /**
+   * @param context used to register and unregister the broadcastReceiver.
```
[the remark this answers is not in the document]
Comment (mklim) [reply packages/connectivity/android/src/main/java/io/flutter/plugins/connectivity/BroadcastReceiverRegistrarImpl.java]: I could see it go either way. This seems fairly straightforward to me but having the explanation about what `context` is used for is also kind of useful.

File: packages/connectivity/android/src/main/java/io/flutter/plugins/connectivity/Connectivity.java [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+    this.wifiManager = wifiManager;
+  }
+
+  @NonNull
+  String checkNetworkType() {
+    return getNetworkType();
+  }
+
+  @Nullable
+  String getWifiName() {
```
[the remark this answers is not in the document]
Comment (mklim) [reply packages/connectivity/android/src/main/java/io/flutter/plugins/connectivity/Connectivity.java]: It's one of those totally subjective style things, like annotating primitive arguments with `/*name=*/`. It's fine either way IMHO. From what I can tell the existing code doesn't use it, so it probably makes more sense to leave it off to be consistent.

File: packages/connectivity/android/src/main/java/io/flutter/plugins/connectivity/BroadcastReceiverRegistrar.java [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+import android.content.BroadcastReceiver;
+import androidx.annotation.NonNull;
+import io.flutter.plugin.common.EventChannel;
+
+/** Responsible for constructing a BroadcastReceiver as well as registering and unregistering it. */
+public interface BroadcastReceiverRegistrar {
+
+  /**
+   * Triggered when it is ready for the BroadcastReceiver to be registered. Register the receiver in
```
[the remark this answers is not in the document]
Comment (mklim) [reply packages/connectivity/android/src/main/java/io/flutter/plugins/connectivity/BroadcastReceiverRegistrar.java]: I think this is a little unclear because it's phrased using passive voice, so it's not obvious who is doing what. https://developers.google.com/style/voice

I'd rephrase the first sentence to `<Object> triggers this when it's ready for the BroadcastReceiver to be registered.` With how it is now I think the worry is that you the developer need to trigger it somehow in some other place. I think the second sentence is clear as-is.

**Comment by cyanglaz** [comment]
@matthew-carroll Updated per Friday's discussion.

File: packages/connectivity/android/src/main/java/io/flutter/plugins/connectivity/Connectivity.java [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+   * @param connectivityManager used to check connectivity information.
+   * @param wifiManager used to check wifi information.
+   */
+  public Connectivity(ConnectivityManager connectivityManager, WifiManager wifiManager) {
+    this.connectivityManager = connectivityManager;
+    this.wifiManager = wifiManager;
+  }
+
+  @NonNull
+  String checkNetworkType() {
```
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply packages/connectivity/android/src/main/java/io/flutter/plugins/connectivity/Connectivity.java]: Any update on this? Does this method need to exist?

File: packages/connectivity/example/android/app/src/main/AndroidManifest.xml [hunk 12]
```diff
@@ -5,11 +5,18 @@
...
-                  android:configChanges="orientation|keyboardHidden|keyboard|screenSize|locale|layoutDirection"
-                  android:hardwareAccelerated="true"
-                  android:windowSoftInputMode="adjustResize">
+            android:launchMode="singleTop"
+            android:theme="@android:style/Theme.Black.NoTitleBar"
+            android:configChanges="orientation|keyboardHidden|keyboard|screenSize|locale|layoutDirection"
+            android:hardwareAccelerated="true"
+            android:windowSoftInputMode="adjustResize">
+        </activity>
+        <activity android:name="dev.flutter.plugins.connectivity_exmaple.MainActivity"
```
Comment (matthew-carroll) [inline packages/connectivity/example/android/app/src/main/AndroidManifest.xml]: FYI, in Java/Android, packages usually don't include underscores. You might choose `plugins.connectivity.example` or `plugins.connectivityexample`.

File: packages/connectivity/android/src/main/java/io/flutter/plugins/connectivity/Connectivity.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (cyanglaz) [reply packages/connectivity/android/src/main/java/io/flutter/plugins/connectivity/Connectivity.java]: No It doesn't. It was copied from the old code. I'll remove this method

File: packages/connectivity/example/android/app/src/main/AndroidManifest.xml (hunk 12 above)
Comment (cyanglaz) [reply packages/connectivity/example/android/app/src/main/AndroidManifest.xml]: using connectivityexample.

**Comment by cyanglaz** [comment]
@collinjackson Could you review the e2e bits in this PR? Although ./gradlew connectedAnroidTest still fails on my local.

File: packages/connectivity/android/src/main/java/io/flutter/plugins/connectivity/ConnectivityMethodChannelHandler.java (hunk 5 above)
Comment (collinjackson) [reply packages/connectivity/android/src/main/java/io/flutter/plugins/connectivity/ConnectivityMethodChannelHandler.java]: I think that Dart e2e integration tests would be acceptable as an alternative to unit tests.
